src/box/box.cc

```diff
@@ -4811,6 +4828,133 @@ box_process_vote(struct ballot *ballot)
 	ballot->registered_replica_uuids_size = i;
 }
 
+/**
+ * The function returns true iff current schema supports
+ * persistent gc consumers.
+ */
+static inline bool
+box_gc_consumer_is_persistent(void)
+{
+	return space_by_id(BOX_GC_CONSUMERS_ID) != NULL;
+}
+
+/**
+ * The function deletes old consumer and sets new one for replica. It is
+ * used on downgrade - we do not have space _gc_consumers then.
+ */
+static int
+box_gc_consumer_set_in_memory(struct replica *replica, const struct vclock *vclock)
+{
+	assert(!box_gc_consumer_is_persistent());
+	/** No in-memory gc consumers for anonymous replicas. */
+	if (replica->anon)
+		return 0;
+	if (replica->gc != NULL)
+		gc_consumer_unregister(replica->gc);
+	replica->gc = gc_consumer_register(vclock, false,
+					   "replica %d", replica->id);
+	return replica->gc != NULL ? 0 : -1;
+}
+
+/**
+ * Sets a persistent gc consumer for replica. Old one is deleted, if any.
+ * Must be used only when schema supports persistent consumers.
+ */
+static int
+box_gc_consumer_set_impl(struct replica *replica, const struct vclock *vclock,
+			 bool with_snap)
+{
+	assert(box_gc_consumer_is_persistent());
+
+	char tuple_buf[VCLOCK_STR_LEN_MAX + UUID_STR_LEN];
+	char *data = tuple_buf;
+	data = mp_encode_array(data, 3);
+	data = mp_encode_str0(data, tt_uuid_str(&replica->uuid));
+	data = mp_encode_vclock_ignore0(data, vclock);
+
+	size_t opts_size = 0;
+	if (with_snap)
+		opts_size++;
+
+	data = mp_encode_map(data, opts_size);
+	if (with_snap) {
+		data = mp_encode_str0(data, "with_snap");
+		data = mp_encode_bool(data, true);
+	}
+	assert((size_t)(data - tuple_buf) < sizeof(tuple_buf));
+
+	return box_replace(BOX_GC_CONSUMERS_ID, tuple_buf, data, NULL);
+}
+
+int
+box_gc_consumer_set(struct replica *replica, const struct vclock *vclock,
+		    bool with_snap)
+{
+	if (!box_gc_consumer_is_persistent())
+		return box_gc_consumer_set_in_memory(replica, vclock);
+
+	return box_gc_consumer_set_impl(replica, vclock, with_snap);
+}
+
+int
+box_gc_consumer_set_force_async(struct replica *replica,
+				const struct vclock *vclock,
+				bool with_snap)
+{
+	if (!box_gc_consumer_is_persistent())
+		return box_gc_consumer_set_in_memory(replica, vclock);
+
+	assert(in_txn() == NULL);
+	struct txn *txn = txn_begin();
+	if (txn == NULL)
+		return -1;
+	txn_set_flags(in_txn(), TXN_FORCE_ASYNC);
+	if (box_gc_consumer_set_impl(replica, vclock, with_snap) != 0) {
+		txn_abort(txn);
+		return -1;
+	}
+	return txn_commit(txn);
+}
+
+int
+box_gc_consumer_update(struct replica *replica, const struct vclock *vclock)
```

Comment:
Why can't we use `box_gc_consumer_set` for this? It seems we don't really need this function

Reply:
```c
/**
 * Update replica gc state in tx thread.
 * Note that there is a case when replica and its consumer were dropped
 * from system spaces, but the relay is not stopped yet - we must not create
 * new gc consumer in this case, because it will become hanging. So the
 * function tries to update gc_consumer - it is not created if it doesn't
 * exist.
 */
static void
tx_gc_advance(struct cmsg *msg)
```